File: sdk/java/src/main/java/io/rlaas/sdk/RlaasClient.java
package io.rlaas.sdk;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.rlaas.sdk.model.CheckRequest;
import io.rlaas.sdk.model.Decision;
import io.rlaas.sdk.model.Policy;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.Map;

public class RlaasClient {
    private final String baseUrl;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public RlaasClient(String baseUrl) {
        this(baseUrl, Duration.ofSeconds(5));
    }

    public RlaasClient(String baseUrl, Duration timeout) {
        this.baseUrl = baseUrl.replaceAll("/+$", "");
        this.httpClient = HttpClient.newBuilder().connectTimeout(timeout).build();
        this.objectMapper = new ObjectMapper();
    }

    public Decision checkLimit(CheckRequest request) throws IOException, InterruptedException {
        String body = objectMapper.writeValueAsString(request);
        HttpResponse<String> response = post("/rlaas/v1/check", body);
        return readJson(response, Decision.class);
    }

    public Map<String, Object> acquire(Map<String, Object> request) throws IOException, InterruptedException {
        HttpResponse<String> response = post("/rlaas/v1/acquire", objectMapper.writeValueAsString(request));
        return readMap(response);
    }

    public Map<String, Object> release(String leaseId) throws IOException, InterruptedException {
        HttpResponse<String> response = post("/rlaas/v1/release", objectMapper.writeValueAsString(Map.of("lease_id", leaseId)));
        return readMap(response);
    }

    public List<Map<String, Object>> listPolicies() throws IOException, InterruptedException {
        HttpResponse<String> response = get("/rlaas/v1/policies");
        return readList(response);
    }

    public Map<String, Object> getPolicy(String policyId) throws IOException, InterruptedException {
        HttpResponse<String> response = get("/rlaas/v1/policies/" + policyId);
        return readMap(response);
    }

    public Map<String, Object> createPolicy(Policy policy) throws IOException, InterruptedException {
        HttpResponse<String> response = post("/rlaas/v1/policies", objectMapper.writeValueAsString(policy));
        return readMap(response);
    }

    public Map<String, Object> updatePolicy(String policyId, Policy policy) throws IOException, InterruptedException {
        HttpResponse<String> response = put("/rlaas/v1/policies/" + policyId, objectMapper.writeValueAsString(policy));
        return readMap(response);
    }

    public void deletePolicy(String policyId) throws IOException, InterruptedException {
        HttpResponse<String> response = delete("/rlaas/v1/policies/" + policyId);
        ensureSuccess(response);
    }

    public Map<String, Object> validatePolicy(Policy policy) throws IOException, InterruptedException {
        HttpResponse<String> response = post("/rlaas/v1/policies/validate", objectMapper.writeValueAsString(policy));
        return readMap(response);
    }

    public Map<String, Object> updateRollout(String policyId, int rolloutPercent) throws IOException, InterruptedException {
        HttpResponse<String> response = post("/rlaas/v1/policies/" + policyId + "/rollout", objectMapper.writeValueAsString(Map.of("rollout_percent", rolloutPercent)));
        return readMap(response);
    }

    public Map<String, Object> rollbackPolicy(String policyId, long version) throws IOException, InterruptedException {
        HttpResponse<String> response = post("/rlaas/v1/policies/" + policyId + "/rollback", objectMapper.writeValueAsString(Map.of("version", version)));
        return readMap(response);
    }

    public List<Map<String, Object>> listPolicyAudit(String policyId) throws IOException, InterruptedException {
        HttpResponse<String> response = get("/rlaas/v1/policies/" + policyId + "/audit");
        return readList(response);
    }

    public List<Map<String, Object>> listPolicyVersions(String policyId) throws IOException, InterruptedException {
        HttpResponse<String> response = get("/rlaas/v1/policies/" + policyId + "/versions");
        return readList(response);
    }

    public Map<String, Object> analyticsSummary(Integer top) throws IOException, InterruptedException {
        String path = "/rlaas/v1/analytics/summary" + (top == null ? "" : "?top=" + top);
        HttpResponse<String> response = get(path);
        return readMap(response);
    }

    private HttpResponse<String> get(String path) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(URI.create(baseUrl + path)).GET().build();
        return httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    }

    private HttpResponse<String> post(String path, String jsonBody) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(URI.create(baseUrl + path))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                .build();
        return httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    }

    private HttpResponse<String> put(String path, String jsonBody) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(URI.create(baseUrl + path))
                .header("Content-Type", "application/json")
                .PUT(HttpRequest.BodyPublishers.ofString(jsonBody))
                .build();
        return httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    }

    private HttpResponse<String> delete(String path) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder(URI.create(baseUrl + path)).DELETE().build();
        return httpClient.send(request, HttpResponse.BodyHandlers.ofString());
    }

    private <T> T readJson(HttpResponse<String> response, Class<T> type) throws IOException {
        ensureSuccess(response);
        return objectMapper.readValue(response.body(), type);
    }

    private Map<String, Object> readMap(HttpResponse<String> response) throws IOException {
        ensureSuccess(response);
        if (response.body() == null || response.body().isBlank()) {
            return Map.of();
        }
        return objectMapper.readValue(response.body(), new TypeReference<Map<String, Object>>() {});
    }

    private List<Map<String, Object>> readList(HttpResponse<String> response) throws IOException {
        ensureSuccess(response);
        if (response.body() == null || response.body().isBlank()) {
            return List.of();
        }
        return objectMapper.readValue(response.body(), new TypeReference<List<Map<String, Object>>>() {});
    }

    private void ensureSuccess(HttpResponse<String> response) {
        int code = response.statusCode();
        if (code >= 400) {
            throw new RuntimeException("RLAAS API error (" + code + "): " + response.body());
        }
    }
}

File: sdk/dotnet/Rlaas.Sdk/RlaasClient.cs
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using Rlaas.Sdk.Models;

namespace Rlaas.Sdk;

public sealed class RlaasClient
{
    private readonly HttpClient _httpClient;
    private readonly string _baseUrl;
    private readonly JsonSerializerOptions _jsonOptions = new(JsonSerializerDefaults.Web)
    {
        PropertyNameCaseInsensitive = true
    };

    public RlaasClient(string baseUrl, HttpClient? httpClient = null)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        _httpClient = httpClient ?? new HttpClient { Timeout = TimeSpan.FromSeconds(5) };
    }

    public Task<Decision> CheckLimitAsync(CheckRequest request, CancellationToken cancellationToken = default) =>
        SendAsync<Decision>(HttpMethod.Post, "/rlaas/v1/check", request, cancellationToken);

    public Task<Dictionary<string, object>> AcquireAsync(Dictionary<string, object> request, CancellationToken cancellationToken = default) =>
        SendAsync<Dictionary<string, object>>(HttpMethod.Post, "/rlaas/v1/acquire", request, cancellationToken);

    public Task<Dictionary<string, object>> ReleaseAsync(string leaseId, CancellationToken cancellationToken = default) =>
        SendAsync<Dictionary<string, object>>(HttpMethod.Post, "/rlaas/v1/release", new Dictionary<string, string> { ["lease_id"] = leaseId }, cancellationToken);

    public Task<List<Dictionary<string, object>>> ListPoliciesAsync(CancellationToken cancellationToken = default) =>
        SendAsync<List<Dictionary<string, object>>>(HttpMethod.Get, "/rlaas/v1/policies", null, cancellationToken);

    public Task<Dictionary<string, object>> GetPolicyAsync(string policyId, CancellationToken cancellationToken = default) =>
        SendAsync<Dictionary<string, object>>(HttpMethod.Get, $"/rlaas/v1/policies/{policyId}", null, cancellationToken);

    public Task<Dictionary<string, object>> CreatePolicyAsync(Policy policy, CancellationToken cancellationToken = default) =>
        SendAsync<Dictionary<string, object>>(HttpMethod.Post, "/rlaas/v1/policies", policy, cancellationToken);

    public Task<Dictionary<string, object>> UpdatePolicyAsync(string policyId, Policy policy, CancellationToken cancellationToken = default) =>
        SendAsync<Dictionary<string, object>>(HttpMethod.Put, $"/rlaas/v1/policies/{policyId}", policy, cancellationToken);

    public async Task DeletePolicyAsync(string policyId, CancellationToken cancellationToken = default)
    {
        using var request = new HttpRequestMessage(HttpMethod.Delete, _baseUrl + $"/rlaas/v1/policies/{policyId}");
        using var response = await _httpClient.SendAsync(request, cancellationToken);
        EnsureSuccess(response);
    }

    public Task<Dictionary<string, object>> ValidatePolicyAsync(Policy policy, CancellationToken cancellationToken = default) =>
        SendAsync<Dictionary<string, object>>(HttpMethod.Post, "/rlaas/v1/policies/validate", policy, cancellationToken);

    public Task<Dictionary<string, object>> UpdateRolloutAsync(string policyId, int rolloutPercent, CancellationToken cancellationToken = default) =>
        SendAsync<Dictionary<string, object>>(HttpMethod.Post, $"/rlaas/v1/policies/{policyId}/rollout", new Dictionary<string, int> { ["rollout_percent"] = rolloutPercent }, cancellationToken);

    public Task<Dictionary<string, object>> RollbackPolicyAsync(string policyId, long version, CancellationToken cancellationToken = default) =>
        SendAsync<Dictionary<string, object>>(HttpMethod.Post, $"/rlaas/v1/policies/{policyId}/rollback", new Dictionary<string, long> { ["version"] = version }, cancellationToken);

    public Task<List<Dictionary<string, object>>> ListPolicyAuditAsync(string policyId, CancellationToken cancellationToken = default) =>
        SendAsync<List<Dictionary<string, object>>>(HttpMethod.Get, $"/rlaas/v1/policies/{policyId}/audit", null, cancellationToken);

    public Task<List<Dictionary<string, object>>> ListPolicyVersionsAsync(string policyId, CancellationToken cancellationToken = default) =>
        SendAsync<List<Dictionary<string, object>>>(HttpMethod.Get, $"/rlaas/v1/policies/{policyId}/versions", null, cancellationToken);

    public Task<Dictionary<string, object>> AnalyticsSummaryAsync(int? top = null, CancellationToken cancellationToken = default)
    {
        var suffix = top.HasValue ? $"?top={top.Value}" : string.Empty;
        return SendAsync<Dictionary<string, object>>(HttpMethod.Get, "/rlaas/v1/analytics/summary" + suffix, null, cancellationToken);
    }

    private async Task<T> SendAsync<T>(HttpMethod method, string path, object? body, CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(method, _baseUrl + path);
        if (body is not null)
        {
            var json = JsonSerializer.Serialize(body, _jsonOptions);
            request.Content = new StringContent(json, Encoding.UTF8, "application/json");
        }

        using var response = await _httpClient.SendAsync(request, cancellationToken);
        EnsureSuccess(response);

        if (response.Content is null)
        {
            return default!;
        }

        var payload = await response.Content.ReadFromJsonAsync<T>(_jsonOptions, cancellationToken);
        return payload!;
    }

    private static void EnsureSuccess(HttpResponseMessage response)
    {
        if ((int)response.StatusCode >= 400)
        {
            throw new InvalidOperationException($"RLAAS API error ({(int)response.StatusCode})");
        }
    }
}
